msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -1064,51 +1060,181 @@ protected void acquireTokenSilent(@NonNull final String[] scopes,
+        final IAccount rootAccount = tokenParameters.getAccount();
+        final MultiTenantAccount multiTenantAccount = (MultiTenantAccount) rootAccount;
+        final String requestAuthority = tokenParameters.getAuthority();
+        final Authority authority = Authority.getAuthorityFromAuthorityUrl(requestAuthority);
 
-        try {
-            final TokenCommand silentTokenCommand = new TokenCommand(
-                    params,
-                    MSALControllerFactory.getAcquireTokenSilentControllers(
-                            mPublicClientConfiguration.getAppContext(),
-                            params.getAuthority(),
-                            mPublicClientConfiguration
-                    ),
-                    callback
+        if (authority instanceof AzureActiveDirectoryB2CAuthority) {
+            // use home account - b2c is not compatible with broker, so no need to construct
+            // the account used in the request...
+            return AccountAdapter.getAccountInternal(
+                    mPublicClientConfiguration.getClientId(),
+                    mPublicClientConfiguration.getOAuth2TokenCache(),
+                    multiTenantAccount.getHomeAccountId(),
+                    multiTenantAccount.getTenantId()
             );
+        } else if (authority instanceof AzureActiveDirectoryAuthority) {
+            final AzureActiveDirectoryAuthority aadAuthority = (AzureActiveDirectoryAuthority) authority;
 
-            ApiDispatcher.submitSilent(silentTokenCommand);
-        } catch (final BaseException exception) {
-            callback.onError(exception);
+            // Although the below call implies the returned value will be a tenantId, this isn't
+            // strictly true: it can also be an alias such as 'common', 'consumers', or
+            // 'organizations'. Additionally, if the developer has used a named tenant it could end
+            // up being something like <tenant_name>.onmicrosoft.com
+            //
+            // If the tenant is a GUID, we need to choose the account/profile that corresponds to it
+            //
+            // If the tenant is named 'common', 'consumers', or 'organizations' we need to use the
+            // home account
+            //
+            // If the account is named like <tenant_name>.onmicrosoft.com we need to query the OpenId
+            // Provider Configuration Metadata in order to get the tenant id. Once we have the
+            // tenant id, we must then select the appropriate home or profile.
+            final String tenantIdNameOrAlias = aadAuthority.getAudience().getTenantId();
+
+            // The AccountRecord we'll use to request a token...
+            final AccountRecord accountRecord = new AccountRecord();
+            accountRecord.setEnvironment(multiTenantAccount.getEnvironment());
+            accountRecord.setHomeAccountId(multiTenantAccount.getHomeAccountId());
+
+            if (isHomeTenantEquivalent(tenantIdNameOrAlias)
```

Comment:
If there's a way to determine that this is not guid (UUID.fromString and catch exception?), we could probably offloading the openID metadata query here and prevent duplicating the logic below.

Reply:
Took your suggestion, applied in a0b7e4ce4c33b99313edb78bfcd9f7351d79822a